### PR TITLE
Fix Issue 18228 - this(this a){} doesn't generate postblit ctor; this(this){} does

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

**Comment by rainers** [comment]
Please also note https://issues.dlang.org/show_bug.cgi?id=12228 which shows that the syntax has been explicitly added at some point, but with very strange limitations. This is not something that is specific to constructors but should be banned in any case, but Walter disagreed in that bug report.

**Review by andralex — APPROVED** [review]

**Comment by rainers** [comment]
> andralex approved these changes 17 hours ago

@andralex Please approve removing using `this` and `super` as a type, too, please. See https://issues.dlang.org/show_bug.cgi?id=12228

I suspect changing this needs a deprecation first, though.

**Review by JinShil — APPROVED** [review]

File: test/fail_compilation/fail18228.d [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail18228.d(11): Error: `this` cannot be used as a parameter type, use `typeof(this)` instead
+fail_compilation/fail18228.d(12): Error: `this` cannot be used as a parameter type, use `typeof(this)` instead
```
Comment (MetaLang) [inline test/fail_compilation/fail18228.d]: This is a [comma splice](https://owl.english.purdue.edu/engagement/2/1/34/). Please change the comma to a semi-colon or period.

**Review by MetaLang — CHANGES_REQUESTED** [review]

File: test/fail_compilation/fail18228.d (hunk 1 above)
Comment (andralex) [reply test/fail_compilation/fail18228.d]: @MetaLang do you have the rights to make changes to PRs?

**Review by andralex — APPROVED** [review]
OK with me. I think we need a deprecation.

**Comment by andralex** [comment]
cc @WalterBright

**Comment by JinShil** [comment]
> OK with me. I think we need a deprecation.

@andralex If that was in response to @rainers comment about removing `this` and `super` as a type, please explicitly state your approval in the issue tracker at https://issues.dlang.org/show_bug.cgi?id=12228

File: test/fail_compilation/fail18228.d (hunk 1 above)
Comment (RazvanN7) [reply test/fail_compilation/fail18228.d]: Done. Thanks!

**Comment by RazvanN7** [comment]
>OK with me. I think we need a deprecation.

Made it a deprecation.

**Review by jacob-carlborg — APPROVED** [review]

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]
A changelog is required as well, describing the problem and what to do to solve it.

**Comment by RazvanN7** [comment]
@jacob-carlborg Done.

File: changelog/deprecate_this_as_param_type.dd [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+whenever it encounters this as a parameter type:
+
+---
+struct A
+{
+    this(this a) {} // Deprecation: `this` cannot be used as a parameter type.
+                    //Use `typeof(this)` instead
+
+    this(int a, this b) {} // ditto
+}
```
Comment (jacob-carlborg) [inline changelog/deprecate_this_as_param_type.dd]: There's no information how to resolve it.

Comment (jacob-carlborg) [reply changelog/deprecate_this_as_param_type.dd]: Well, I guess it's mentioned in the deprecation message which is included above.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
Do we need a spec update? How are we dealing with deprecations and spec updates?

**Pull request merged**